### DISCUSSION TITLE
Using linear history when merging PRs

### DISCUSSION THREAD

**Discussion by m-lindgren** [UniMath/UniMath#1678] [General]
I'm suggesting we enable [enforce linear history](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/defining-the-mergeability-of-pull-requests/about-protected-branches#require-linear-history) as a branch protection. The only difference this will have is that when merging a PR we will not have the option to create a merge commit, but instead we may choose between `squash and merge` and `rebase`.

The benefits of using `squash and merge` is most evident when a PR consists of many commits that  are relatively small, but dependent on each other. For example a PR often contains commits of the kind `fix spelling` followed by `oops forgot one` -- and this is great, I encourage everyone to do small commits and commit often during development, and I absolutely do not mind PR's containing such commits at all. But merging PR's into `master` directly first of all creates an ugly merge-commit, and second all of those commits gets interspersed in the commit history making it impossible to connect a commit / change to a specific PR, effectively making the git log chaotic. Another example is when changes in master has been merged into a PR branch several times, so that the final commit history will be a sequence of meaningless merge commits. 

We (persons who can merge PRs into master) can avoid this by using `squash and merge` (instead of `create merge commit`) to take all the commits in a PR and squash them into one and append it to the master branch.

When squash-merging the title of the commit will by default have the title that the PR has, and the number of the PR. The text of the commit can be set to the message in the PR, but here the person merging it to master must look at it and maybe clean it up, or replace it entirely to give it a good short description (and remove the `Co-authored by` line!).

An example where I've done this is in the last few PR:s that I merged by @nmvdw : https://github.com/UniMath/UniMath/commits/master - the only editing I had to do was add the `*` in front of each line.

If we do this consistently we will have a very clean `git log` with no merge commits taking up space and no small commits polluting the history, and when looking at git-blame every line will be connected to a specific PR. We might even be able to produce a Changelog associated to each release. The extra work one has to do before merging a PR is very minimal, essentially just cleaning up the description or shortening it so that it looks good in a `git log`. Removing details from the description is not a problem since the original description is still available by following the link in the commit title to the PR.

What do you think about this? Should we try it out for a while?

**Comment by benediktahrens** [comment]
Thanks for explaining this in detail! I would be happy to try it out.

**Reply by m-lindgren** [reply]
Great! I've enabled it in the master branch for now.

**Comment by rmatthes** [comment]
@m-lindgren Can you please also tell us about the remaining option `rebase`? How does it take care of all other cases?

I think we will be more and more dependent on GitHub, but we are also being well served by it. The downside is that the mere Git repo will not be sufficient to develop UniMath further efficiently.

**Reply by m-lindgren** [reply]
A rebase takes all the commits and uses `git rebase` to place the commits in a PR after the latest commit in `master`, so the history will be linear. It will potentially contain a lot of minor commits like `fix spelling`.
